Feature/KAS-3081 Pub v3 Details: Info and remark-panel

mdv-professional opened the pull request:
[Jira](https://kanselarij.atlassian.net/browse/KAS-3081)

De layout is iets anders dan in usertest: twee rijen leken me overzichtelijker.
Ik heb zoveel mogelijk proberen te behouden van de sidebar/new-publications-modal i.v.m. error-messages,...

Commit:
(busy) add remark panel
KAS-3081 pub: v3: details: info and remark

Commit:
add remark panel
KAS-3081 pub: v3: details: info + remark

Commit:
info tooltip conforming styleguide in publication remark panel
KAS-3081 pub: v3: details: info + remark

Commit:
info tooltip as separate component
KAS-3081 pub: v3: details: info + remark

Commit:
correct test selector name
KAS-3081 pub: v3: details: info + remark

Commit:
info panel urgency of publication details
KAS-3081 pub: v3: details: info + remark

Commit:
info panel: publication number (raw version) in publications/details
KAS-3081 pub: v3: details: info + remark

Commit:
info panel: change sizes to allow error message in publications/details
KAS-3081 pub: v3: details: info + remark

Commit:
Auk::Label keep regular label style when error
KAS-3081 pub: v3: details: info + remark
@brenner-company

Commit:
info panel: publication number: validate and save in publications/details
KAS-3081 pub: v3: details: info + remark

Commit:
info panel: numac numbers (raw version) in publications/details
KAS-3081 pub: v3: details: info + remark

Commit:
Merge branch 'development' into feature/KAS-3081-pub-v3-details--info-and-remark

Commit:
Merge branch 'feature/KAS-3024-pub-feedback--update-columns' into feature/KAS-3081-pub-v3-details--info-and-remark

Commit:
info panel: numac number in publications/details
KAS-3081 pub: v3: details: info + remark

Commit:
info panel: dates (raw version) in publications/details
KAS-3081 pub: v3: details: info + remark

Commit:
info panel: dates (raw version) in publications/details
KAS-3081 pub: v3: details: info + remark

Commit:
info panel: dates (raw version) in publications/details
KAS-3081 pub: v3: details: info + remark

Commit:
info panel: dates: Model dirty checking does not work as expected
KAS-3081 pub: v3: details: info + remark

Commit:
info panel: re-layout
KAS-3081 pub: v3: details: info + remark

Commit:
info panel: publication number: refactor validation + shorten error message
KAS-3081 pub: v3: details: info + remark

Commit:
info panel: urgent
KAS-3081 pub: v3: details: info + remark

Commit:
(lint-fix) info panel
KAS-3081 pub: v3: details: info + remark

Commit:
info panel: remove icon to agenda-item-treatment
KAS-3081 pub: v3: details: info + remark

Commit:
(lint-fix) remark panel
KAS-3081 pub: v3: details: info + remark

Commit:
(code-style) use let instead of var
KAS-3081 pub: v3: details: info + remark

Commit:
(comment)
KAS-3081 pub: v3: details: info + remark

Commit:
(translations) update "numac-number"
KAS-3081 pub: v3: details: info + remark

Commit:
(translations) remove unused key "no-numac-number-yet"
KAS-3081 pub: v3: details: info + remark

Commit:
(translations) "Uiterste" > "Limiet"
KAS-3081 pub: v3: details: info + remark

Commit:
(style) improve spacing
KAS-3081 pub: v3: details: info + remark

Commit:
add publication number tooltip
KAS-3081 pub: v3: details: info + remark

Commit:
use fields for date input
KAS-3081 pub: v3: details: info + remark

Commit:
import ember-concurrency decorators from 'ember-concurrency'
KAS-3081 pub: v3: details: info + remark

Commit:
(style) improve alignment of publication number input fields
KAS-3081 pub: v3: details: info + remark

Commit:
info panel: publication number: improve error check
KAS-3081 pub: v3: details: info + remark

Commit:
(translations) tab and panel titles in publications/details
KAS-3081 pub: v3: details: info + remark

Commit:
(fix) info panel: publication number: error check
KAS-3081 pub: v3: details: info + remark

Commit:
(cleanup) remove unused variable
KAS-3081 pub: v3: details: info + remark

Commit:
(translations) uniform keys for titles
KAS-3081 pub: v3: details: info + remark

Commit:
(move) info-panel > publication-case-info-panel
reason: consistency
KAS-3081 pub: v3: details: info + remark

Commit:
(rename) info-panel > publication-case-info-panel
reason: consistency
KAS-3081 pub: v3: details: info + remark

Commit:
(refactor) change edit-mode layout
reason: consistency
KAS-3081 pub: v3: details: info + remark

Commit:
(refactor) remove component argument isViaCouncilOfMinisters
KAS-3081 pub: v3: details: info + remark

Commit:
include subcases and agenda-item-treatment in publication detail request
KAS-3081 pub: v3: details: info + remark

Commit:
info panel: add "Datum verstreken" voor Limiet publicatie
KAS-3081 pub: v3: details: info + remark

Commit:
info panel: save logic
KAS-3081 pub: v3: details: info + remark

Commit:
Merge branch 'development' into feature/KAS-3081-pub-v3-details--info-and-remark

Commit:
temporary async getter workaround
KAS-3081 pub: v3: details: info + remark

Commit:
(rename) update data-test attributes
KAS-3081 pub: v3: details: info + remark

Commit:
simplification of component properties
KAS-3081 pub: v3: details: info + remark

erikap committed:
Merge branch 'development' into feature/KAS-3081-pub-v3-details--info-and-remark

Commit:
Merge branch 'development' into feature/KAS-3081-pub-v3-details--info-and-remark

Commit:
(code-style) still prefer const
KAS-3081 pub: v3: details: info + remark

Commit:
(code-style) still prefer const
KAS-3081 pub: v3: details: info + remark

erikap commented on app/components/publications/publication/case/info/publication-case-info-panel.js:
```diff
@@ -0,0 +1,271 @@
+    this.publicationDueDate = await this.getPublicationDueDate(publicationFlow);
+  }
+
+  async getIsViaCouncilOfMinisters(publicationFlow) {
+    const _case = await publicationFlow.case;
+    const subcases = await _case.subcases;
+    return !!subcases.length;
+  }
+
+  async getDecisionDate(publicationFlow) {
+    const agendaItemTreatment = await publicationFlow.agendaItemTreatment;
+    return agendaItemTreatment.startDate;
+  }
+
+  async getPublicationDueDate(publicationFlow) {
+    const publicationSubcase = await publicationFlow.publicationSubcase;
+    return publicationSubcase.dueDate;
+  }
```
Ik snap dat helper methods soms nuttig zijn om code duplication te vermijden of te zorgen dat methods niet té lang worden, maar anderzijds verhoogt alles opsplitsen in methods de leesbaarheid ook niet altijd omdat je te veel moet springen tussen methods.
Aangezien deze methods, voor zover ik zie, maar op 1 plaats gebruikt worden zou je in `initFields` toch ook bijvoorbeeld kunnen schrijven
```
this.decisionDate = (await publicationFlow.publicationSubcase).dueDate
```

mdv-professional replied:
Voor de `get...Date`-methods begrijp ik het.
De `getIsViaCouncilOfMinisters`-methode zou ik naar de `PublicationService` verhuizen. Ik weet dat ge hier niet altijd voorstander van zijt. Ik herinner me van het begin van de opstart van publicaties, dat dit iets was dat ik steeds terug moest opzoeken. Vandaar dat ik hem ergens graat ter referentie zou zetten.

erikap replied:
Voor die `getIsViaCouncilOfMinisters` snap ik dat wel, want dat is inderdaad minder triviale logica. Als je hem naar daar verhuist, probeer dan zeker in de rest van de code eens te kijken waar je die method nog kan herbruiken.

erikap commented on app/components/publications/publication/case/info/publication-case-info-panel.js:
```diff
@@ -0,0 +1,271 @@
+  }
+
+  // separate method to prevent ember-concurrency from saving only partially
+  async performSave(publicationFlow) {
```
Ben je in deze task niet heel veel boekhouding die aangeboden wordt door ember-data zelf aan het doen?
Ik weet dat ember-data niet geschikt is voor dirty-tracking op relaties, maar hier gebeurt het voornamelijk op gewone attributes terwijl je daarvoor [`hasDirtyAttributes`](https://api.emberjs.com/ember-data/4.1/classes/Model/properties/hasDirtyAttributes?anchor=hasDirtyAttributes) en [`rollbackAttributes`](https://api.emberjs.com/ember-data/4.1/classes/Model/methods/rollbackAttributes?anchor=rollbackAttributes) hebt.

mdv-professional replied:
Nadat ik het zo had gedaan stelde ik me eerder de vraag of de checks uberhaupt nodig zijn. Misschien is het, kwestie van het simpel te houden best om altijd alles opnieuw te saven?

erikap replied:
Op dit moment gebeurt die check nergens in de applicatie (en er zijn plaatsen waar dit veel meer meerwaarde zou hebben dan hier). We zouden liefst een generieke oplossing hebben die dit algemeen oplost voor de hele applicatie zonder dat je daar overal aan moet denken, maar dat is ver buiten de scope van deze story :slightly_smiling_face: 
 
Ter info, in de backend wordt er ook gewerkt aan het optimaliseren van mu-authorization/delta-handler zodanig dat er alleen delta-messages gestuurd worden voor de triples die effectief gewijzigd worden. Dat gaat de overhead ook al deels verminderen.

Voor mij mag je de checks hier dus zeker weglaten om de code eenvoudiger te maken.

erikap commented on app/components/publications/publication/case/info/publication-case-info-panel.js:
```diff
@@ -0,0 +1,271 @@
+
+  @tracked isViaCouncilOfMinisters;
+
+  @tracked isUrgent;
+
+  @tracked publicationNumberErrorKey;
+  @tracked publicationNumber;
+  @tracked publicationNumberSuffix;
+
+  @tracked numacNumbers;
+  numacNumbersToDelete;
+
+  @tracked decisionDate;
+  @tracked openingDate;
+  @tracked publicationDueDate;
```
Is het nodig om van al die properties een lokale `@tracked` property bij te houden? Maken we het daarmee niet complexer dan nodig en is het niet makkelijker om gewoon rechtstreeks op de ember-data models te werken? Voor sommige zaken, zoals die numacNummers, zal het inderdaad wel handiger zijn, maar voor voor bijvoorbeeld de dates en `isUrgent` lijkt het mij een onnodige extra laag complexiteit die wordt toegevoegd.

Edit: ik heb me pas later in de review beseft dat de properties van `this.args.publicationFlow` inderdaad gecached moeten worden omdat ze anders misschien mee opgeslagen zullen worden bij `save` in het Opmerkingen-panel. Voor die cases zou ik dat dan expliciet documenteren dat dat de reden is.

erikap commented on app/components/publications/publication/case/info/publication-case-info-panel.js:
```diff
@@ -0,0 +1,271 @@
+  }
+
+  @action
+  onChangeIsUrgent(ev) {
```
Aangezien dit geen action handler is die als argument meegegeven wordt, de `on` best vermijden.

erikap commented on app/components/publications/publication/case/info/publication-case-info-panel.js:
```diff
@@ -0,0 +1,271 @@
+
+  @action
+  onChangeIsUrgent(ev) {
+    const isUrgent = ev.target.checked;
+    this.isUrgent = isUrgent;
```
In lijn met de vorige comment: kan hier niet gewoon rechtstreeks met het `urgencyLevel` record gewerkt worden ipv de extra indirectie van `isUrgent`?

mdv-professional replied:
Die `isUrgent`-boolean zou ik behouden. De extra indirectie is er toch, want de checkbox verwacht een boolean als model.
Maar misschien vindt ge de toevoegingen van `getIsUrgent` en `getUrgencyLevel` aan de `PublicationService` geen meerwaarde?

erikap replied:
Aangezien de checkbox-component data-down-actions-up is, kan je, als je het `urgencyLevel` ter beschikking hebt, aan de checkbox `@checked={{urgencyLevel.isUrgent}}` meegeven. En in de `onChange`-handler stel je dan het juiste level in op basis van de boolean die je binnenkrijgt.
De help methods in de publication-service vind ik inderdaad wat omslachtig; zeker omdat die dan nog eens in een andere file staan.

erikap commented on app/components/publications/publication/case/remark/remark-panel.js:
```diff
@@ -0,0 +1,36 @@
+
+  @tracked remark;
+
+  constructor() {
+    super(...arguments);
+  }
```
Overbodig?

erikap commented on app/components/publications/publication/case/remark/remark-panel.js:
```diff
@@ -0,0 +1,36 @@
+  putInEditMode() {
+    const publicationFlow = this.args.publicationFlow;
+    this.isInEditMode = true;
+    this.remark = publicationFlow.remark;
```
Best documenteren waarom hier met een lokale copy van `remark` gewerkt wordt ipv rechtstreeks met `publicationFlow.remark`

erikap commented on app/components/publications/publication/case/remark/remark-panel.js:
```diff
@@ -0,0 +1,36 @@
+  *save() {
+    const publicationFlow = this.args.publicationFlow;
+    publicationFlow.remark = this.remark;
+    // no try-catch: don't exit if save didn't work
```
Als de save niet gewerkt heeft, moet `publicationFlow.remark` wel gerollbacked worden. Anders wordt dat misschien mee opgeslagen bij het opslaan van een andere panel.

mdv-professional replied:
Eigenlijk hebt ge daar gelijk, maar is dat niet te veel uitzonderingen opvangen?

erikap replied:
Ja, waarschijnlijk wel. Want dan zouden we dat ook op het andere panel moeten doen en daar is de rollback-strategie al veel complexer omdat er meerdere saves/promises zijn. Negeer deze comment dus maar.

erikap reviewed (changes requested):
De layout zou ik, tenzij overlegd met de designers, behouden zoals in usertest. Zij hebben immers met de gebruikers overlegd.

Op veel plaatsen wordt `let` gebruikt waar `const` mogelijk is.

Het feit dat properties van publication-flow in verschillende panels, met elk een eigen save-button, bewerkt kunnen worden, is een complexiteit die ik initieel niet gezien had. Ik zou zeker goed documenteren dat dat de reden is waarom op sommige plaatsen local copies van properties gebruikt worden.

erikap commented:
@mdv-professional Het verwijderen van de velden, die nu geediteerd kunnen worden via de panels, uit de sidebar ontbreekt nog in deze PR

Commit:
info panel: remove get*Date methods
KAS-3081 pub: v3: details: info + remark

Commit:
(comment) add comment for publication dates
KAS-3081 pub: v3: details: info + remark

Commit:
(cleanup) remove useless constructor in remark panel
KAS-3081 pub: v3: details: info + remark

Commit:
(comment) clarify use of copied properties
KAS-3081 pub: v3: details: info + remark

Commit:
remove duplicate sidebar fields
KAS-3081 pub: v3: details: info + remark

Commit:
info panel: hide publication due date for withdrawn publication-flow
KAS-3081 pub: v3: details: info + remark

Commit:
(lint-fix)
KAS-3081 pub: v3: details: info + remark

Commit:
info panel: fields in single row
KAS-3081 pub: v3: details: info + remark

Commit:
(lint-fix)
KAS-3081 pub: v3: details: info + remark

Commit:
(code-style) info panel: component action handler without "on"
KAS-3081 pub: v3: details: info + remark

Commit:
(cleanup)
KAS-3081 pub: v3: details: info + remark

Commit:
(refactor) move urgency-level methods from PublicationService to PublicationCaseInfoPanel controller
KAS-3081 pub: v3: details: info + remark

Commit:
(comment) info panel: initFields method
KAS-3081 pub: v3: details: info + remark

Commit:
(refactor) move getIsViaCouncilOfMinisters method to PublicationService
KAS-3081 pub: v3: details: info + remark

Commit:
(refactor) replace PublicationService.getIsViaCouncilOfMinisters(...) by PublicationFlow.isViaCouncilOfMinisters
KAS-3081 pub: v3: details: info + remark

Commit:
info panel: always save all models
KAS-3081 pub: v3: details: info + remark

Commit:
Merge branch 'development' into feature/KAS-3081-pub-v3-details--info-and-remark

erikap committed:
Merge branch 'feature/KAS-3081-pub-v3-details--info-and-remark' of github.com:kanselarij-vlaanderen/kaleidos-frontend into feature/KAS-3081-pub-v3-details--info-and-remark

erikap committed:
Merge branch 'development' into feature/KAS-3081-pub-v3-details--info-and-remark

erikap commented:
Ik heb alle vorige comments gemarkeerd als resolved om het terug wat overzichtelijker te krijgen. Er zijn nu nog allerlei commits en changes gebeurd, maar die hebben de vorige comments maar deels opgelost en nieuwe dingen geïntroduceerd (vb. de async getter `isViaCouncilOfMinisters` is nu naar het publication-flow model verhuisd).

Ik zou even een stap terugnemen van je huidige implementatie, eens kijken naar de components `publications/publication/publication-activities/publications-info-panel`, `publications/publication/translations/translations-info-panel.js` en `publications/publication/proofs/proof-info-panel.js` die recent in development gemerged zijn en de panels in deze PR op een gelijkaardig manier proberen op te bouwen:
- rechtstreeks op Ember Data models werken waar mogelijk
- wijzigingen ongedaan maken adhv `rollbackAttributes()` en reloads van relaties (zoals [hier](https://github.com/kanselarij-vlaanderen/frontend-kaleidos/blob/development/app/components/publications/publication/publication-activities/publications-info-panel.js#L33) bijvoorbeeld). De edge cases waarbij we dan ook attributes van andere panels mogelijks rollbacken, gaan we negeren.
- analoge naamgeving voor properties/acties (vb. `isEditing` vs `isInEditMode`)
- async get-functions proberen vermijden

De `isViaCouncilOfMinisters` zou ik terugzetten zoals het was in de routes/controllers. Je kan een helper-method op de publication-service toevoegen, zoals je eerst gedaan had, om duplicate code te vermijden, maar een async get-function op het model is een pattern waar we van weg willen.

Commit:
Merge branch 'development' into feature/KAS-3081-pub-v3-details--info-and-remark

Commit:
revert translation key change
KAS-3081 pub: v3: details: info + remark

Commit:
align naming
KAS-3081 pub: v3: details: info + remark

Commit:
consistence: use isUrgent property of urgency-level in template + use rollbackAttributes in closeEditingPanel
KAS-3081 pub: v3: details: info + remark

Commit:
consistence: use model properties in template: publication number
KAS-3081 pub: v3: details: info + remark

Commit:
fix: ignore current publication-flow when checking whether publication number is already taken
KAS-3081 pub: v3: details: info + remark

Commit:
consistence: use model properties in template: numac numbers
KAS-3081 pub: v3: details: info + remark

Commit:
consistence: use model properties in template: dates
KAS-3081 pub: v3: details: info + remark

Commit:
move isViaCouncilOfMinisters back to PublicationService
KAS-3081 pub: v3: details: info + remark

Commit:
(eslint)
KAS-3081 pub: v3: details: info + remark

Commit:
Merge branch 'feature/KAS-3081-pub-v3-details--info-and-remark' of github.com:kanselarij-vlaanderen/frontend-kaleidos into feature/KAS-3081-pub-v3-details--info-and-remark

ghost commented:
Deze issues zouden in orde moeten zijn.

Commit:
Merge branch 'development' into feature/KAS-3081-pub-v3-details--info-and-remark

Commit:
fix incorrect merge
KAS-3081 pub: v3: details: info + remark

Commit:
clearer code
KAS-3082 pub: v3: details: decision info

erikap committed:
Inverse negative if-clauses for readability

erikap committed:
Consistent comment formatting in template

erikap committed:
Update panel fields layout to match design and other info panels

erikap committed:
Don't obfuscate publication flow is coming from component args

erikap committed:
Use isUrgent flag of urgency-level model to find correct urgency level

erikap committed:
Reuse structured identifier update logic of publication-sidebar

Not going to change the error behaviour in this branch

erikap committed:
Fix rollback of numac numbers on cancel

Locally created numac-numbers must be removed to make sure the
list is up-to-date

erikap committed:
Add missing await in initFields

erikap committed:
Update action names on incription-panel to match other publication panels

erikap committed:
Disable save on inscription panel if input is not valid

erikap committed:
Inverse negative if-clauses for readability

erikap committed:
Fix linting

erikap committed:
Disable publication cypress tests

erikap committed:
Disable publication tests

erikap committed:
Use error messages on publication number input instead of toasts as originally
proposed

erikap committed:
Sync validation errors on new-publication-modal and publication-case-info-panel

erikap committed:
Use auk/form-help-text to display short title input error

erikap committed:
Disable another publication test
